flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/SqlTryCastFunction.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.functions.sql;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.planner.calcite.FlinkTypeFactory;
+import org.apache.flink.table.planner.functions.casting.CastRuleProvider;
+import org.apache.flink.table.types.logical.LogicalType;
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlCallBinding;
+import org.apache.calcite.sql.SqlFunctionCategory;
+import org.apache.calcite.sql.SqlIntervalQualifier;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlOperandCountRange;
+import org.apache.calcite.sql.SqlOperatorBinding;
+import org.apache.calcite.sql.SqlSyntax;
+import org.apache.calcite.sql.SqlWriter;
+import org.apache.calcite.sql.fun.SqlCastFunction;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+
+import static org.apache.flink.table.functions.BuiltInFunctionDefinition.DEFAULT_VERSION;
+
+/**
+ * This class implements the {@code TRY_CAST} built-in, essentially delegating all the method
+ * invocations, whenever is possible, to Calcite's {@link SqlCastFunction}.
+ */
+@Internal
+public class SqlTryCastFunction extends BuiltInSqlFunction {
+
+    /**
+     * Note that this constructor is mimicking as much as possible the constructor of Calcite's
+     * {@link SqlCastFunction}.
+     */
+    SqlTryCastFunction() {
+        super(
+                "TRY_CAST",
+                DEFAULT_VERSION,
+                SqlKind.OTHER_FUNCTION,
+                null,
+                SqlStdOperatorTable.CAST
+                        .getOperandTypeInference(), // From Calcite's SqlCastFunction
+                null,
```

Comment:
We should try to have a complete copy of `CAST`. E.g. `CAST` defines `InferTypes.FIRST_KNOWN` also the monotonicity property can be adopted.

Reply:
And in fact we're adopting it, without copying the actual implementation, but rather simply using the getters to get monotonicity and operand type inference. it's essentially no different than directly using `InferTypes.FIRST_KNOWN` here, with the difference that this potentially is one less pain to fix when upgrading to the next calcite version